1235-Dashboard-add radio buttons w moving average breakdown by wk and mo, normalize by population per 10k people and update labels

piotrsan opened the pull request:
Fixes #{issue number here}

  - [ ] Up to date with `dev` branch
  - [ ] Branch name follows [guidelines](https://github.com/hackforla/311-data/blob/master/GETTING_STARTED.md#feature-branching)
  - [ ] All PR Status checks are successful
  - [ ] Peer reviewed and approved

Any questions? See the [getting started guide](https://github.com/hackforla/311-data/blob/master/GETTING_STARTED.md)

nichhk commented on server/dash/dashboards/neighborhood.py:
```diff
@@ -39,27 +42,80 @@ def populate_options():
         )
     return values
 
+print(populate_options())
```
If we want to log this in prod, let's make it more detailed, e.g.,

`print("neighborhood.populate_options: output is ", populate_options())`

piotrsan replied:
thanks for catching this, this was for me to understand the code when I started

piotrsan replied:
I will get rid of this

nichhk commented on server/dash/dashboards/neighborhood.py:
```diff
@@ -39,27 +42,80 @@ def populate_options():
+        placeholder="Select a neighborhood",
+        options=populate_options()
+    ),
+    dcc.Dropdown(
```
Mentioned this in person, but it would be nice to make this dropdown a component of the first graph since it only applies to that graph.

piotrsan replied:
I made a new pull request with radio options now found below the graph.

nichhk commented on server/dash/dashboards/neighborhood.py:
```diff
@@ -39,27 +42,80 @@ def populate_options():
+    total_sum_df["week_n"] = total_sum_df['created_date'].apply(lambda x: str(x.isocalendar()[1]))
+    total_sum_df["month_n"] = total_sum_df['created_date'].dt.month
+
+    total_sum_df['nc_avg_Mo'] = total_sum_df.groupby(['month_n'])['counts'].transform('mean')
```
Also mentioned this in person, but I think a moving average would be more meaningful.

piotrsan replied:
Thanks for suggesting that. I changed both the specific council and the 99 council avg to a moving avg. The only issue is that with only two months worth of data we can only compute those for a partial time frame.

nichhk commented on server/dash/dashboards/neighborhood.py:
```diff
@@ -39,27 +42,80 @@ def populate_options():
+
     total_sum_df = df.groupby(['created_date']).agg('sum').reset_index()
+
     total_sum_df["nc_avg"] = total_sum_df["counts"] / 99
```
While you're here, could you create a global variable for this at the top of the file? Following the GLOBAL_CONSTANT_NAME convention. https://google.github.io/styleguide/pyguide.html#316-naming

piotrsan replied:
I have change this code. I am not clear what would benefit from being a global variable here. The new pull request has simpler code. I think it could probably be simplified yet using a long table instead of the current format.

nichhk replied:
Apologies, I was talking specifically about `99`. Magic numbers are discouraged. See this stackoverflow [answer](https://stackoverflow.com/a/47902). It's in Java, but the same idea applies to all languages.

piotrsan replied:
That's a great convention, thanks for pointing this out. I was actually confused at first as to what 99 was. Partially this is why this ticket was first generated as it was not clear what the graph represented. Hopefully, now it will be more clear.

nichhk commented on server/dash/dashboards/neighborhood.py:
```diff
@@ -39,27 +42,80 @@ def populate_options():
+    total_sum_df["week_n"] = total_sum_df['created_date'].apply(lambda x: str(x.isocalendar()[1]))
+    total_sum_df["month_n"] = total_sum_df['created_date'].dt.month
+
+    total_sum_df['nc_avg_Mo'] = total_sum_df.groupby(['month_n'])['counts'].transform('mean')
```
nit: let's not mix in uppercase here. 'nc_avg_mo'

piotrsan replied:
Thanks. Do we have instructions on best coding practices and naming conventions for this project?

nichhk replied:
I don't think any style guide has been followed so far, but the [Google Python Style Guide](https://google.github.io/styleguide/pyguide.html) is preferred.

I'm not sure if this particular style convention would be mentioned in that guide above, but in general, Python sticks to `snake_case` for variable names (and I think dict keys would fall under that) and `CamelCase` for class names. Mixing `snake_case` and `CamelCase` (e.g., `nc_avg_Mo`) is not recommended.

piotrsan replied:
Thanks

nichhk commented on server/dash/dashboards/neighborhood.py:
```diff
@@ -39,27 +42,80 @@ def populate_options():
-        labels=LABELS,
         title="Comparison trend for " + selected_council
     )
+    newnames = {"counts_x": "Count of 311 Requests", "nc_avg_Mo": "Monthly Council Average", "nc_avg_wk": "Weekly Council Average", "nc_avg": "Daily Council Average"}
```
We can pull out the nc_avg* keys into a separate dict and use that both for renaming here and the dropdown above. That way, if we want to change the names, we only need to do it in one place.

piotrsan replied:
I have changed the way the code works. Now it might be closer to what you suggested.

nichhk commented on server/dash/dashboards/neighborhood.py:
```diff
@@ -39,27 +42,80 @@ def populate_options():
     total_sum_df["nc_avg"] = total_sum_df["counts"] / 99
-    merged_df = neighborhood_sum_df.merge(total_sum_df["nc_avg"].to_frame(), left_index=True, right_index=True)  # noqa
 
+    total_sum_df["week_n"] = total_sum_df['created_date'].apply(lambda x: str(x.isocalendar()[1]))
```
Could you leave some comments explaining what week_n and month_n are?

piotrsan replied:
These are no longer needed.

nichhk commented on server/dash/dashboards/neighborhood.py:
```diff
@@ -18,12 +18,15 @@
 df = pd.read_json(API_HOST + query_string)
 print(" * Dataframe has been loaded")
 
+df["created_date"] = pd.to_datetime(df['created_date'])
```
Could you leave a comment explaining why we need to do this?

piotrsan replied:
This is no longer needed.

piotrsan replied:
got rid of it

nichhk commented on server/dash/dashboards/neighborhood.py:
```diff
@@ -39,27 +42,80 @@ def populate_options():
 )
-def update_figure(selected_council):
-
+def update_figure(selected_council, select_timeframe):
```
nit: selected_timeframe

piotrsan replied:
Chnaged this.

nichhk commented on server/dash/dashboards/neighborhood.py:
```diff
@@ -39,27 +42,80 @@ def populate_options():
 )
-def update_figure(selected_council):
-
+def update_figure(selected_council, select_timeframe):
```
Could you create a docstring here explaining what this method does? https://google.github.io/styleguide/pyguide.html#383-functions-and-methods

piotrsan replied:
I gave this a shot, let me know if this works.

nichhk reviewed (changes requested):
Thanks for working on this Piero!

Please update the title and description of this PR to reflect the changes that you've made.

Most of my comments are related to readability, and I know some of the issues already existed beforehand. We are working on improving the readability of our code, and we're asking everyone to improve where they can. Sorry for the extra requests, but I think it will be valuable for future contributions!

nichhk commented on server/dash/dashboards/neighborhood.py:
```diff
@@ -14,16 +14,20 @@
 from design import DISCRETE_COLORS
 from design import LABELS
 
+NEIGHBORHOOD_COUNCILS_NUMBER=99 #Number of nc in Los Angeles as of 06.2022
```
Please add spaces around the '='. Are you using an auto-formatter? It should take care of issues like this.

Please read https://google.github.io/styleguide/pyguide.html#385-block-and-inline-comments and https://google.github.io/styleguide/pyguide.html#386-punctuation-spelling-and-grammar re: the comment.

piotrsan replied:
I added an auto formatter (black). Thanks!

piotrsan replied:
> Please add ".idea/*" to ".gitignore". We don't want local IDE configs to be included in the repo

Hi Nich,
I saw that 6 files that I did not modify willingly were added to the PR. Do you know what those are and how would I got about adding those to gitignore?
Thanks!

nichhk replied:
You can add ".idea/*" to the [gitignore](https://github.com/hackforla/311-data/blob/dev/.gitignore) file.

nichhk commented on server/dash/dashboards/neighborhood.py:
```diff
@@ -43,35 +47,67 @@ def populate_options():
     apply_figure_style(fig)
 
     return fig
+"""Creates comparison graph
+
+
+    Args:
+        selected_council: The selected council from the drop down menu
+        selected_timeframe: radio button selection of time window for computing moving average
+
+
+    Returns:
+        Graph that compares the selected moving average between the chosen council and the moving average of the avg of all
+        the 99 neighborhood councils 
+
+"""
```
Please move this to be right under `def update_figure...`

Please add periods at the end of each sentence and include datatypes for the Args and Returns.

piotrsan replied:
I changed it, let me know if the data type are correct

nichhk commented on server/dash/dashboards/neighborhood.py:
```diff
@@ -43,35 +47,67 @@ def populate_options():
-    total_sum_df["nc_avg"] = total_sum_df["counts"] / 99
-    merged_df = neighborhood_sum_df.merge(total_sum_df["nc_avg"].to_frame(), left_index=True, right_index=True)  # noqa
 
+    total_sum_df['nc_ma'] = total_sum_df.counts.rolling(selected_timeframe, center=True).mean()/NEIGHBORHOOD_COUNCILS_NUMBER
```
Add spaces around `/`

piotrsan replied:
formatter moved it to next line, is that fine?

nichhk commented on server/dash/dashboards/neighborhood.py:
```diff
@@ -107,23 +143,50 @@ def update_council_figure(selected_council):
-        config=CONFIG_OPTIONS
-    )
+layout = html.Div(children=[
+    
```
newlines in this block aren't needed.

piotrsan replied:
formatter took care of these

nichhk reviewed (changes requested):
Thanks for the updates!

Can you include screenshots of the updated dashboard?

Please run an auto-formatter. There are some simple issues with spacing. Also, quote usage is inconsistent. Please use double-quotes throughout the file.

Please add ".idea/*" to ".gitignore". We don't want local IDE configs to be included in the repo

piotrsan commented on server/dash/dashboards/neighborhood.py:
```diff
@@ -43,35 +47,67 @@ def populate_options():
-    total_sum_df["nc_avg"] = total_sum_df["counts"] / 99
-    merged_df = neighborhood_sum_df.merge(total_sum_df["nc_avg"].to_frame(), left_index=True, right_index=True)  # noqa
 
+    total_sum_df['nc_ma'] = total_sum_df.counts.rolling(selected_timeframe, center=True).mean()/NEIGHBORHOOD_COUNCILS_NUMBER
```
Can you give me an example for this? I am not sure I quite get what you mean.

nichhk replied:
You can make a constant like:

`NC_MOVING_AVERAGE_KEY = "nc_ma"  # Key for neighborhood council moving average in dataframe.`

Similar to how you made a constant for the number of neighborhood councils.

piotrsan replied:
Hi Nich, I implemented this, but I am not sure it is correct. I don't completely understand why I would create a variable name to a column name, other than having a place where I can write a comment about what the variable does. Am I getting this right?

piotrsan commented:
Here is the current screenshot of the first part of the Dashboard:
<img width="1664" alt="image" src="https://user-images.githubusercontent.com/15342705/173993210-1b12da0d-203f-4182-98b6-7561fee21eff.png">

Here how it looks with 7day MA.
<img width="1652" alt="image" src="https://user-images.githubusercontent.com/15342705/173993331-cccd3fb8-09a8-42a7-bc8e-c92f7fc362b2.png">

nichhk commented on server/dash/dashboards/neighborhood.py:
```diff
@@ -14,13 +14,15 @@
 from design import DISCRETE_COLORS
 from design import LABELS
 
+NEIGHBORHOOD_COUNCILS_NUMBER = 99  # Number of nc in Los Angeles as of 06.2022
```
nit: period at end of comment

nichhk commented on server/dash/dashboards/neighborhood.py:
```diff
@@ -30,59 +32,97 @@
 )
-def update_figure(selected_council):
+def update_figure(selected_council, selected_timeframe):
+    """Creates comparison graph
```
nit: period after "graph".

nichhk commented on server/dash/dashboards/neighborhood.py:
```diff
@@ -30,59 +32,97 @@
+
+    Args:
+        selected_council:
+          A string representing the council selected from the drop down menu.
```
Did the formatter put "A string..." on a new line? The arg and the description are supposed to be on the same line. Ditto for "Returns" section.

piotrsan replied:
I think it did. I changed it to reflect your comment

nichhk commented on server/dash/dashboards/neighborhood.py:
```diff
@@ -30,59 +32,97 @@
+    )
+    newnames = {
+        "select_nc_ma": selected_council,
+        "nc_ma": "99 Neighborhood Councils Average",
```
I think "Neighborhood Council Average" is a better name here.

nichhk commented on server/dash/dashboards/neighborhood.py:
```diff
@@ -30,59 +32,97 @@
-        labels=LABELS,
-        title="Comparison trend for " + selected_council
+        labels={"created_date": "311 Request Date", "value": "Total Requests"},
+        title=selected_council + " vs 99 Neighborhood Councils 311 Requests Average",
```
same: "Neighborhood Council Average"

nichhk replied:
We should probably make a constant for this as well so that we don't need to worry about updating it in multiple places.

piotrsan replied:
what do you think we should make a constant for?

piotrsan replied:
gotcha

nichhk commented on server/dash/dashboards/neighborhood.py:
```diff
@@ -30,59 +32,97 @@
         color_discrete_sequence=DISCRETE_COLORS,
-        labels=LABELS,
-        title="Comparison trend for " + selected_council
+        labels={"created_date": "311 Request Date", "value": "Total Requests"},
```
Let's omit "311" from here.

nichhk commented on server/dash/dashboards/neighborhood.py:
```diff
@@ -30,59 +32,97 @@
+        labels={"created_date": "311 Request Date", "value": "Total Requests"},
+        title=selected_council + " vs 99 Neighborhood Councils 311 Requests Average",
+    )
+    newnames = {
```
nit: new_names

nichhk commented on server/dash/dashboards/neighborhood.py:
```diff
@@ -91,39 +131,67 @@ def update_council_figure(selected_council):
+                    children=[
+                        html.Label("Moving Average Time Window:"),
+                    ],
+                    style={"width": "14%", "display": "inline-block"},
```
Please leave a comment in the code describing how you chose 14% here. Ditto for the 15% below.

I'm not sure if this will have issues on mobile devices. We can check it after this code change is checked in.

piotrsan replied:
This was by trial and error. This needs to be reworked with UX

nichhk replied:
Please leave a comment to that effect, so that when people read this code, they can understand what's stable and what is planned to be changed.

nichhk commented on server/dash/dashboards/neighborhood.py:
```diff
@@ -30,59 +32,97 @@
+          This is selected with radio buttons at the bottom of the graph.
+
+    Returns:
+        fig:
```
For "Returns", we don't need to specify the name of the returned variable. See the [example](https://google.github.io/styleguide/pyguide.html#383-functions-and-methods) in the style guide.

nichhk reviewed (changes requested):
Just a few more style changes. Looking good!

nichhk commented:
It's totally fine for this PR, but in the future, let's split this into two PRs: one for the name change, and one for the moving average change.

piotrsan commented:
Thanks Nich.

I have implemented some normalization, should that be made into a new ull request?

nichhk commented:
Yes, please make a new PR. But before you make the PR, please open an issue explaining why we need normalization

nichhk reviewed (changes requested):
Thanks Piero! I see that you resolved several comments, but it doesn't seem like they were fixed. Did you forget to push some commits?

nichhk commented on server/dash/dashboards/neighborhood.py:
```diff
@@ -37,63 +39,90 @@ def populate_options():
+    )
+    NC_POP_MOVING_AVERAGE_KEY = "nc_ma_pop"  # Key for neighborhood council moving average adjusted to per 10,000 people.
+    SELECT_NC_MOVING_AVERAGE_KEY = "select_nc_ma"  # Key for selected neighborhood council moving average in dataframe.
+    SELECT_NC_POP_MOVING_AVERAGE_KEY = "select_nc_ma_pop"  # Key for selected neighborhood council moving average in dataframe adjusted to per 10,000 people.
```
Please move these constants to the top of the file where other constants are defined.

nichhk commented on server/dash/dashboards/neighborhood.py:
```diff
@@ -37,63 +39,90 @@ def populate_options():
-        labels={"created_date": "311 Request Date", "value": "Total Requests"},
-        title=selected_council + " vs 99 Neighborhood Councils 311 Requests Average",
+        labels={"created_date": "Request Date", "value": "Total Requests"},
+        title=selected_council + " vs Neighborhood Councils Average",
```
So I was suggesting to make a constant for "Neighborhood Councils Average" because it's used multiple times in this file. In the case that we want to change this string (e.g., through usability testing, we find that this name is confusing or another name is better), we will only need to change it in one place.

piotrsan replied:
I simplified the title instead and removed the unnecessary variables

nichhk commented on server/dash/dashboards/neighborhood.py:
```diff
@@ -37,63 +39,90 @@ def populate_options():
+    total_sum_df[NC_POP_MOVING_AVERAGE_KEY] = (
+        total_sum_df.counts.rolling(selected_timeframe, center=True).mean()
+        / TOT_POP
+        * 10000
```
Please make a  constant for 10000.

nichhk commented on server/dash/dashboards/neighborhood.py:
```diff
@@ -37,63 +39,90 @@ def populate_options():
+    new_names = {
+        SELECT_NC_MOVING_AVERAGE_KEY: selected_council,
+        NC_MOVING_AVERAGE_KEY: "Neighborhood Councils Average",
+        SELECT_NC_POP_MOVING_AVERAGE_KEY: selected_council + "council adjusted",
```
Let's make it clear to the user that we're calculating the number of 311 requests per 10,000 people.

piotrsan replied:
Changed the title of the graph

nichhk commented on server/dash/NC_pop_2020.csv:
```diff
@@ -0,0 +1,100 @@
+council_name,sum_pop20,area_sqmi,pop_density
```
can we rename `sum_pop20` to `population`?

nichhk commented on .gitignore:
```diff
@@ -26,3 +26,6 @@ config.js
 dist
+
+#311
+.idea/*
```
Sorry, my suggestion didn't work. I think you can change this to `.idea/` (i.e., get rid of the *). Also, I don't think the comment above ("#311") is useful.

nichhk reviewed (changes requested):
Thanks Piero! Just a few more comments.

piotrsan commented on server/dash/dashboards/neighborhood.py:
```diff
@@ -14,13 +14,15 @@
 
-NEIGHBORHOOD_COUNCILS_NUMBER = 99  # Number of nc in Los Angeles as of 06.2022
+NEIGHBORHOOD_COUNCILS_NUMBER = 99  # Number of nc in Los Angeles as of 06.2022.
+TOT_POP = 3814700
```
done thanks

piotrsan commented:
Hi Nich,

thanks for the comments. I implemented all changes and restricted this particular dashboard for showcasing how a specific council compares to the average (this has to be done by normalizing by population for the comparison to be meaningful). I removed the raw counts, which can be visualized in other dashboards or in a different graph if needed.

piotrsan commented:
The current visualization is done using population data from NC_pop_2020.csv which was generated from arcGS public data analysis (it is not clear how the analysis is done by the app). Anupriya will provide the final numbers through her careful analysis, which can then be integrated into the dashboard.

Skydodle commented:
Closed without merge due to outdated PR & irrelevant to current working branch `main`.
